feat: implement secondsOfInactivityBeforeIdling and secondsOfRunBefor…

dkwon17 opened the pull request:
…eIdling CR config

Signed-off-by: David Kwon <dakwon@redhat.com>

<!-- Please review the following before submitting a PR:
che-operator Development Guide: https://github.com/eclipse-che/che-operator/#development
-->

### What does this PR do?
Part of https://github.com/eclipse/che/issues/21390.

This PR adds `spec.devEnvironments.secondsOfInactivityBeforeIdling` and `spec.devEnvironments.secondsOfRunBeforeIdling` config in the CheCluster CR:
<img width="1213" alt="image" src="https://user-images.githubusercontent.com/83611742/178339993-d5dae0b9-3400-4dd9-8256-26fff789b4f3.png">

This PR reads the above config to create a configmap named `che-idle-settings` when a user namespace is reconciled. The configmap is created so that the data can be mounted to the devworkspace containers. This PR for che-machine-exec: https://github.com/eclipse-che/che-machine-exec/pull/208 reads these environment variables. Here is an example of the generated config map:
![image](https://user-images.githubusercontent.com/83611742/178340100-a91208a5-38d7-4b0d-a163-8befcaf544b4.png)

### What issues does this PR fix or reference?
<!-- Please include any related issue from eclipse che repository (or from another issue tracker).
     Include link to other pull requests like documentation PR from [the docs repo](https://github.com/eclipse/che-docs)
-->
Part of https://github.com/eclipse/che/issues/21390.

### How to test this PR?
<!-- Please explain for example :
  - The test platform (openshift, kubernetes, minikube, docker-desktop, etc)
  - steps to reproduce
 -->


To test on an OpenShift cluster:
1. Create and install Che operator with this catalog source file:
```yaml
apiVersion:  operators.coreos.com/v1alpha1
kind:         CatalogSource
metadata:
  name:         eclipse-che-custom-catalog
  namespace:    openshift-operators
spec:
  image:       quay.io/dkwon17/eclipse-che-openshift-opm-catalog:next
  sourceType:  grpc
  updateStrategy:
    registryPoll:
      interval: 5m
```

with this che-operator image: `quay.io/dkwon17/che-operator:idle-cr`

2. Create a CheCluster CR with `spec.devEnvironments.secondsOfInactivityBeforeIdling` set to "300" and `spec.devEnvironments.secondsOfRunBeforeIdling` set to "360".

3. Open the dashboard to provision a user namespace.

4. In the user namespace verify that the `che-idle-settings` configmap in the user namespace is created, with the following data:
```
SECONDS_OF_DW_INACTIVITY_BEFORE_IDLING: "300"
SECONDS_OF_DW_RUN_BEFORE_IDLING: "360"
```
5. Edit the  `spec.devEnvironments.secondsOfInactivityBeforeIdling` and `spec.devEnvironments.secondsOfRunBeforeIdling` properties in the CheCluster CR to different values and verify that the changes exist in the `che-idle-settings` configmap.

### PR Checklist

[As the author of this Pull Request I made sure that:](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#pull-request-template-and-its-checklist)

- [x] [The Eclipse Contributor Agreement is valid](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#the-eclipse-contributor-agreement-is-valid)
- [x] [Code produced is complete](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#code-produced-is-complete)
- [x] [Code builds without errors](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#code-builds-without-errors)
- [x] [Tests are covering the bugfix](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#tests-are-covering-the-bugfix)
- [x] [Custom resource definition file is up to date](https://github.com/eclipse-che/che-operator/blob/main/README.md#updating-custom-resource-definition-file)
- [x] [OLM bundles are up to date](https://github.com/eclipse-che/che-operator/blob/main/README.md#update-olm-bundle)
- [ ] [The repository devfile is up to date and works](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#the-repository-devfile-is-up-to-date-and-works)
- [x] [Sections `What issues does this PR fix or reference` and `How to test this PR` completed](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#sections-what-issues-does-this-pr-fix-or-reference-and-how-to-test-this-pr-completed)
- [ ] [Relevant user documentation updated](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#relevant-contributing-documentation-updated)
- [ ] [Relevant contributing documentation updated](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#relevant-contributing-documentation-updated)
- [ ] [CI/CD changes implemented, documented and communicated](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#cicd-changes-implemented-documented-and-communicated)

### Reviewers

Reviewers, please comment how you tested the PR when approving it.

tolusha commented on api/v2/checluster_types.go:
```diff
@@ -90,6 +90,16 @@ type CheClusterDevEnvironments struct {
+	// This timeout is the duration after which a workspace will be idled if there is no activity.
+	// If this value is not provided, workspaces will not idle.
+	// +optional
+	SecondsOfInactivityBeforeIdling string `json:"secondsOfInactivityBeforeIdling,omitempty"`
```
Do we have any default value? if so, pls add the following annotation `// +kubebuilder:default:=`
I am thinking about some sort of validation `// +kubebuilder:validation:Pattern=^\d+[ywdhms]$`
WDYT?

dkwon17 replied:
Hi @l0rd , should the default value for workspace idling be set to `900` (= 15 minutes) ? Then, if the admin wants to remove idling the admin would need to set it to maybe `-1`, how does that sound?

Currently, this PR does has no workspace idling by default

l0rd replied:
Good point @tolusha, we should have default values and they should be set to 1800 (30 mins) and -1 respectively

dkwon17 replied:
Sounds great, fixed https://github.com/eclipse-che/che-operator/pull/1443#issuecomment-1182466989

tolusha commented:
@dkwon17 
You have to add the fields to CheCluster API v1 as well [1] and implement conversion like here for instance [2]

[1] https://github.com/eclipse-che/che-operator/blob/3830ac4e28983dbe928d0777f63c87bed43feb4b/api/v1/checluster_types.go#L867
[2] https://github.com/eclipse-che/che-operator/pull/1395/files

l0rd commented:
@tolusha yes, having the fields in the `DevEnvironments` is what I suggested [in the issue description](https://github.com/eclipse/che/issues/21390#issue-1232423102).

tolusha commented on api/v2/checluster_types.go:
```diff
@@ -90,6 +90,16 @@ type CheClusterDevEnvironments struct {
+	// If this value is not provided, workspaces will not idle.
+	// +optional
+	SecondsOfInactivityBeforeIdling string `json:"secondsOfInactivityBeforeIdling,omitempty"`
+	// Run timeout for workspaces, for example: 15m, 30m, 1h.
```
BWT, There is some inconsistency between field name (starts with `Seconds...`) and description (we mention time interval like `15m`).  If everything about seconds, then we need to mention duration in minutes or hours? 
Should it be like
`SecondsOfRunBeforeIdling *int32 `json:"secondsOfRunBeforeIdling,omitempty"` ?

dkwon17 replied:
Oops, yes my mistake. It should be `SecondsOfRunBeforeIdling *int32`, and the comments must be updated

dkwon17 commented:
Thanks all for the feedback

> You have to add the fields to CheCluster API v1 as well [1] and implement conversion like here for instance [2]

Fixed

I updated this PR to have default values of `900` and `-1` to `spec.devEnvironments.secondsOfInactivityBeforeIdling` and `spec.devEnvironments.secondsOfRunBeforeIdling` respectively:

![image](https://user-images.githubusercontent.com/83611742/178583257-9966a5f6-ce53-426b-a8a4-65f938ae339d.png)

Here is the `che-idle-settings` configmap that is created for a user namespace: 
![image](https://user-images.githubusercontent.com/83611742/178583289-86bb3fa8-6d74-4991-b1dc-91bf41302bf3.png)

tolusha commented on api/v1/checluster_types.go:
```diff
@@ -724,6 +724,16 @@ type CheClusterSpecDevWorkspace struct {
+	// Idle timeout for workspaces in seconds.
+	// This timeout is the duration after which a workspace will be idled if there is no activity.
+	// To disable workspace idling due to inactivity, set this value to -1.
+	// +kubebuilder:default:=900
```
Pls. mention the defaults here as well
https://github.com/eclipse-che/che-operator/blob/885d623258450b6c942d7396c4f3c17ff05b0f2e/api/v2/checluster_types.go#L40

tolusha commented on api/v1/checluster_types.go:
```diff
@@ -724,6 +724,16 @@ type CheClusterSpecDevWorkspace struct {
+	// Run timeout for workspaces in seconds.
+	// This timeout is the maximum duration a workspace runs.
+	// To disable workspace run timeout, set this value to -1.
+	// +kubebuilder:default:=-1
```
Pls. mention the defaults here as well
https://github.com/eclipse-che/che-operator/blob/885d623258450b6c942d7396c4f3c17ff05b0f2e/api/v2/checluster_types.go#L40

tolusha commented on api/v2/checluster_types.go:
```diff
@@ -90,6 +90,16 @@ type CheClusterDevEnvironments struct {
+	// Idle timeout for workspaces in seconds.
+	// This timeout is the duration after which a workspace will be idled if there is no activity.
+	// To disable workspace idling due to inactivity, set this value to -1.
+	// +kubebuilder:default:=900
```
Pls. mention the defaults here as well
https://github.com/eclipse-che/che-operator/blob/885d623258450b6c942d7396c4f3c17ff05b0f2e/api/v2/checluster_types.go#L40

tolusha commented on api/v2/checluster_types.go:
```diff
@@ -90,6 +90,16 @@ type CheClusterDevEnvironments struct {
+	// Run timeout for workspaces in seconds.
+	// This timeout is the maximum duration a workspace runs.
+	// To disable workspace run timeout, set this value to -1.
+	// +kubebuilder:default:=-1
```
Pls. mention the defaults here as well
https://github.com/eclipse-che/che-operator/blob/885d623258450b6c942d7396c4f3c17ff05b0f2e/api/v2/checluster_types.go#L40

tolusha commented on controllers/usernamespace/controller.go:
```diff
@@ -448,6 +455,50 @@ func (r *CheUserNamespaceReconciler) reconcileProxySettings(ctx context.Context,
+func (r *CheUserNamespaceReconciler) reconcileIdleSettings(ctx context.Context, targetNs string, checluster *chev2.CheCluster, deployContext *chetypes.DeployContext) error {
+
+	noIdle := pointer.Int32Ptr(-1)
+	if checluster.Spec.DevEnvironments.SecondsOfInactivityBeforeIdling == noIdle && checluster.Spec.DevEnvironments.SecondsOfRunBeforeIdling == noIdle {
```
I think we have to compare with `nil` otherwise defaults value `-1` won't be propagated to Config Map.

tolusha commented on controllers/usernamespace/controller.go:
```diff
@@ -448,6 +455,50 @@ func (r *CheUserNamespaceReconciler) reconcileProxySettings(ctx context.Context,
+	data := map[string]string{}
+
+	if checluster.Spec.DevEnvironments.SecondsOfInactivityBeforeIdling != noIdle {
+		data["SECONDS_OF_DW_INACTIVITY_BEFORE_IDLING"] = strconv.Itoa(int(*checluster.Spec.DevEnvironments.SecondsOfInactivityBeforeIdling))
```
I think it is better to use `strconv.FormatInt(int64(<VALUE>), 10)` not to lost precision

tolusha commented:
@dkwon17 
It is good to merge.

dkwon17 commented:
Thank you everyone, I will squash and merge
